source/mainnet/smart-contracts/guides/migrate-contracts.rst

```diff
@@ -0,0 +1,164 @@
+But |concordium-smart-contract-testing|_ uses the exact same state implementation as the nodes do and it is therefore possible to specify the concrete type, |StateApi|_, as the default.
+
+Until the |test_infrastructure|_ module is completely removed, the libraries will still support it, and thus the generic parameter ``S`` must still be present.
+
+To migrate an advanced state type you must change the following:
+
+- Set ``StateApi`` as the default type for the generic parameters that must implement |HasStateApi|_ (typically named ``S``)
+- Remove the generic type parameter for the |HasStateApi|_ type where you use your state type, e.g., ``MyState`` instead of ``MyState<S>``
+
+Here is an example of the change you must make:
+
+.. code-block:: rust
+   :emphasize-lines: 15, 23
+
+   /// Before
+   struct MyState<S> {
+       my_map: StateMap<AccountAddress, TokenCount, S>,
+   }
+
+   #[init(contract = "contract_before")]
+   fn init_before<S: HasStateApi>(
+       ctx: &impl HasInitContext,
+       state_builder: &mut StateBuilder<S>,
+   ) -> InitResult<MyState<S>> {
+       Ok(MyState{ my_map: state_builder.new_map() })
+   }
+
+   /// After
+   struct MyState<S = StateApi> {
+       my_map: StateMap<AccountAddress, TokenCount, S>,
+   }
+
+   #[init(contract = "contract_before")]
+   fn init_before(
+       ctx: &InitContext,
+       state_builder: &mut StateBuilder,
+   ) -> InitResult<MyState> {
+       Ok(MyState{ my_map: state_builder.new_map() })
+   }
+
+Reference material for migrating types and tests
+================================================
+
+The examples above show how to migrate most contracts, but for advanced contracts, there may be more types to migrate.
+The list below shows how to achieve that.
+It also includes types already described above:
+
+- ``&impl HasInitContext`` becomes ``&InitContext``
+- ``&impl HasReceiveContext`` becomes ``&ReceiveContext``
+- ``&mut StateBuilder<S>`` becomes ``&mut StateBuilder``
+- ``&impl HasHost<MyState, StateApiType = S>`` becomes ``&Host<MyState>``
+- ``&mut impl HasHost<MyState, StateApiType = S>`` becomes ``&mut Host<MyState>``
+- When using the ``low_level`` attribute:
+
+  - On inits: ``&mut impl HasStateApi`` becomes ``&mut StateApi``
+  - On receives: ``&mut impl Host<S>`` becomes ``&mut LowLevelHost``
+
+- ``struct MyState<S> { my_map: StateMap<_,_, S> }`` becomes ``struct MyState<S = StateApi> { my_map: StateMap<_, _, S> }``
+- ``&impl HasCryptoPrimitives`` becomes ``&CryptoPrimitives``
+- ``&impl HasChainMetadata`` becomes ``&ChainMetadata``
+- ``&mut impl HasLogger`` becomes ``&mut Logger``
+
+To migrate your tests, read the how-to guide :ref:`integration-test-contract`.
+You can also refer to the `commit <https://github.com/Concordium/concordium-rust-smart-contracts/commit/d26aabae4b0714412cec4347d7ca5b09c8994c0e>`_ where the example contracts were rewritten.
```

Comment:
Just make sure we don't rebase or otherwise destroy this commit.

Would it make more sense to link to the PR instead?

Reply:
The PR seemed like the safer choice.